NetKet/Hamiltonian/abstract_hamiltonian.hpp

```diff
@@ -19,8 +19,24 @@
 #include <complex>
 #include <vector>
 
+#include "Hilbert/hilbert.hpp"
+
 namespace netket {
 
+// clang-format off
```

Comment:
This is the last thing I am not entirely convinced about! I mean, not the clang-format off (btw, why do you have this here?) but rather the MatrixElement struct. 
The thing is that I would expect a MatrixElement object to carry information both about v and v', not only v'! Maybe we can add a span pointing to v? Would it make sense?

Reply:
Not sure it makes sense to point to `v`. In places where this struct makes sense to appear, there should always be a known `v'`.

We also could just as well directly pass the matrix element value and two `span`s to the callback. I found the struct a bit easier to read, though. Maybe the name implies a generality that is not not there (and not needed). Would calling it something like `Connection` help?

> the clang-format off (btw, why do you have this here?)

Hm, `clang-format` did add an awkward line break in the struct (due to the comments).

Reply:
Alright agreed, ```MatrixConnector``` then is a better concept?

Reply:
In this way in the future we can maybe also have that ```FindConn``` returns a vector of ```MatrixConnector```

Reply:
> In this way in the future we can maybe also have that FindConn returns a vector of MatrixConnector

That would not work well, I think. Since the `span`s do not own the memory they refer to, one needs to be careful with their lifetime. Hence the comment here:
https://github.com/netket/netket/blob/baa7c0d6a29ebd2194b31b2c475fab3848f3d4a0/NetKet/Hamiltonian/abstract_hamiltonian.hpp#L77-L80
That the parameter in `ForEachConn` is only usable inside the callback seems reasonably safe to me. However, I would expect to be able to pass the return value of `FindConn` around safely, without worrying about its lifetime. Thus, a `vector` seems to be the better choice here.

Therefore, I do not see a good way to unify these at the moment. I also propose `ConnectorRef` as a name. The `Ref` part makes it clear that one has to think about the lifetime if this object is passed around.